File: practice/algorithms/implementation/jumping_on_the_clouds/jumping_on_the_clouds.cpp
#include <vector>
#include <iostream>

using namespace std;

int solve(const vector<int> & i_c, int i_n, int i_pos)
{
    if (i_pos < i_n - 1)
    {
        int next_pos = i_pos + 1;
        if (i_pos < i_n - 2 && i_c[i_pos + 2] == 0)
        {
            next_pos++;
        }
        return 1 + solve(i_c, i_n, next_pos);
    }
    
    return 0;
}

int main()
{
    int n;
    cin >> n;
    vector<int> c(n);
    for (int pos = 0; pos < n; ++pos)
    {
       cin >> c[pos];
    }
    
    cout << solve(c, n, 0) << endl;
    
    return 0;
}

File: practice/algorithms/implementation/jumping_on_the_clouds/jumping_on_the_clouds.hs
module Main where

import qualified Data.Vector as V

solve :: V.Vector Int -> Int -> Int -> Int
solve c n pos
    | pos >= n - 1 = 0
    | otherwise    = solve c n next_pos + 1
    where
        next_pos = if pos < n - 2 && (c V.! (pos + 2) == 0) then pos + 2 else pos + 1

main :: IO ()
main = do
    n  <- readLn :: IO Int
    c' <- getLine
    let c = map (read :: String -> Int) $ words c'
    print $ solve (V.fromList c) n 0

